### ISSUE TITLE
Support file downloads

### ISSUE THREAD

**Issue by pocmo** [mozilla-mobile/focus-android#14]
**Theme**: _Default Browser_

**User Story**

> I discovered content of a page that I want to revisit later (save to my device)

**Feature**

> Support for downloads (file downloads, PDF, execs)

**Background**

Currently just nothing happens when clicking a link that should start a download.

This should be fairly easy to implement using WebView (I guess), but could be more complicated with Gecko. Downloads are not part of the GeckoView MVP. However we might be able to work around that with some custom JavaScript.

Maybe we just want to intercept the download and let Android's download manager perform the task (privacy concerns?).

**Tasks**
- [x] Implement [DownloadListener](https://developer.android.com/reference/android/webkit/WebView.html#setDownloadListener(android.webkit.DownloadListener))
- [x] UX mocks / flow
- [x] Show dialog to let the user confirm the download (see UI mock)
- [x] Show snackbar once download is finished
- [x] Request runtime permission to write to external storage
- [x] Forward Download to Android's Download Manager
- [x] Add telemetry probe

**UI mock**

<img width="1532" alt="screen shot 2017-07-06 at 8 43 59 pm" src="https://user-images.githubusercontent.com/51007/27905544-f7023f36-628b-11e7-9331-7fbaf11afdd2.png">

**Comment by pocmo** [comment]
@antlam I implemented downloads and they will ask for the storage permission (if needed). However I assume we want to show some kind of UI before starting the download?

We know the following things at the time of the download and could display it:
* URL from which we are downloading
* File name (This is either defined by the server or generated by us)
* MIME type (Optional)
* Size of file (Optional)

**Comment by antlam** [comment]
Gonna leave this one to @brampitoyo !

**Comment by pocmo** [comment]
Btw.: Downloads are already implemented (without UI, except that we may ask for a permission) - but not enabled in builds as we decided to not ship it in v1.

**Comment by danilodorgam** [comment]
[Request runtime permission to write to external storage](https://developer.android.com/training/permissions/requesting.html)

Follows documentation about the request.

If no one implements this function, I can try to do this later.

**Comment by pocmo** [comment]
> If no one implements this function, I can try to do this later.

That's already implemented!

**Comment by brampitoyo** [comment]
First take on the download dialogue:

![download dialog - mdpi](https://cloud.githubusercontent.com/assets/51007/26570449/6629004a-455c-11e7-9668-aa480c17700d.png)

When user taps on a link that would start a download, we’d show a standard Android permissions dialogue (I suppose we can show a light theme one, but I’ve chosen a dark theme one for continuity).

Another thinking I had was to label the button “Continue” instead of “Download”. When you tap that button, we actually won’t start downloading straight away. Instead, we’ll pass it on to Android’s Download Manager, that will perform the task of choosing file destination, actually starting the download, etc.

What do you think?

**Comment by bbinto** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Support for downloads (file downloads, PDF, execs)](https://trello.com/c/LGziUxWU/26-support-for-downloads-file-downloads-pdf-execs)

**Comment by brampitoyo** [comment]
Here’s that permission dialogue in context of the download flow.

1. Tap on a download link.
2. We show a permission dialogue. You don’t have to show it every time. Tap “Download” to proceed.
3. We pass the job over to Android’s Download Manager

<img width="1136" alt="screen shot 2017-06-15 at 5 18 52 pm" src="https://user-images.githubusercontent.com/51007/27167888-cedf2e40-51ee-11e7-8c83-aebaae7e82e8.png">

**Comment by pocmo** [comment]
Note that we might need to request a permission to save files. In this case we'd show two dialogs in a row:

![permissions](https://user-images.githubusercontent.com/89638/27539374-9e2a6364-5a7c-11e7-9dab-6fa9cb89f9a7.png)

I like the wording in your dialog. What I'm not sure about is the "Always remember my decision" checkbox. What decision did the user make? Isn't this more like an informal dialog? What is expected to happen if the user presses "remember" and "cancel"?

I wonder if we could merge screen 2 and 3 into one?

**Comment by brampitoyo** [comment]
> Note that we might need to request a permission to save files. In this case we'd show two dialogs in a row.

Showing two dialogues in a row doesn’t sound ideal. On the other hand, we also don’t get to control what to show on Android’s default permissions dialogue.

Maybe this revised flow can be a decent solution:

* Don’t write our own dialogue. Use Android system’s.
* Still let Android’s Download Manager handle folder and saving
* After download has proceeded, show a snackbar with warning, along with an “Undo” button (alternative copy: “Cancel”)

<img width="1526" alt="screen shot 2017-06-27 at 11 58 55 am" src="https://user-images.githubusercontent.com/51007/27566325-41d2406e-5b30-11e7-88bb-7ab9cd76ee0b.png">

**Comment by brampitoyo** [comment]
Here’s a comparison of how different Android Browsers handle their download confirmation dialogues in Private Browsing Mode:

https://docs.google.com/a/mozilla.com/presentation/d/1iUsLkW4aOLDwbZmHFKcJFtRBz3gmTQVZUoFFLXcZ3Us/edit

This is the flow that Firefox for Android has implemented:

1. On download start, show snackbar: “Download started…”
2. On download completed, show snackbar: “document.pdf downloaded &nbsp; [OPEN]”

We can take this, along with other implementations, into consideration.

**Comment by pocmo** [comment]
@brampitoyo Great list. Thank you for compiling it. I personally like the simplicity of Opera Mini. I wonder whether "download location" and "file name" is something a core requirement (for a privacy browser)? We at least want to warn that downloads are not erased automatically, right?

I landed the current version, that has no UI except the notification that Android's Download manager shows.

> Don’t write our own dialogue. Use Android system’s.

I couldn't find anything in the Android documentation about how this could be triggered. Where did you see this screen?

**Comment by antlam** [comment]
It sounds like we need some content/copy help here! @mheubusch maybe you have some thoughts?

**Comment by brampitoyo** [comment]
@pocmo, I’ve hashed out this issue with @antlam, and we’ve agreed to augment the current flow with two small things below.

@mheubusch, we’d love it if you could take a look at these two elements below and see whether our copy looks alright.

1. Download confirmation dialogue

> **Download file**
> <file-icon> <file-name.ext>
> [Cancel] &nbsp; [Download]
> Downloaded files will not be deleted when you erase Firefox Focus history.

2. Snackbar for when the download has completed downloading

> <file-name.ext> finished &nbsp; &nbsp; &nbsp; [Open]

<img width="1532" alt="screen shot 2017-07-06 at 8 43 59 pm" src="https://user-images.githubusercontent.com/51007/27905544-f7023f36-628b-11e7-9331-7fbaf11afdd2.png">

Unlike Firefox for Android, no snackbar will show up on download start.

**Comment by pocmo** [comment]
Looks great. I updated the description of the issue with the final mock and tasks.

**Comment by mheubusch** [comment]
@brampitoyo @antlam Copy in last set of comps looks good!

**Comment by ekager** [comment]
@brampitoyo is there an existing resource for the file icon in the dialog? I didn't see it in the folder of Android resources where the blocking icon is.

**Comment by brampitoyo** [comment]
@ekager wrote:
> @brampitoyo is there an existing resource for the file icon in the dialog?

We’re using Android’s Material icon. You can find the SVG here: https://storage.googleapis.com/material-icons/external-assets/v4/icons/svg/ic_insert_drive_file_white_24px.svg

**Comment by ekager** [comment]
Hey UX - @antlam @brampitoyo 
When we show the completed download snackbar with the option to open, if there is no activity found to open the type of file the user has downloaded, we shouldn't just show nothing when "Open" is pressed. Should we not show the "Open" action in the first place if no apps can be found to open the intent, or should we show an error toast/dialog that no activity to handle this intent could be found when the user presses "Open"? The error toast or dialog would need strings.

**Comment by brampitoyo** [comment]
I just talked to @ekager over on Slack. When no app is available to handle a file type, we think that we should keep the same message (“filename.ext finished”) and remove the “Open” button.

As we get to reuse the snackbar message, we get more consistency in the UI that way.

**Comment by mcomella** [comment]
@ekager Can we close this now that #885 has landed?
